sql-stream-to-stream-join-test/src/main/java/com/hazelcast/jet/tests/sql/SqlStreamToStreamJoinSoakTest.java

```diff
@@ -0,0 +1,221 @@
+
+        Util.sleepMillis(30_000L);
+
+        SqlResult sqlResult = sqlService.execute("SELECT * FROM v1 JOIN v2 ON v1.event_time=v2.event_time");
+
+        Iterator<SqlRow> iterator = sqlResult.iterator();
+        while (System.currentTimeMillis() - startTime < durationInMillis) {
+            SqlRow sqlRow = iterator.next();
+
+            // l_event       r_event
+            // time(0) | 0 | time(0) | 0  <-- joint row
+
+            // checking only event_tick to simplify our destiny.
+            Long leftRowIndex = sqlRow.getObject(1);
+            Long rightRowIndex = sqlRow.getObject(3);
+            AbstractSoakTest.assertEquals(leftRowIndex, rightRowIndex);
+
+            currentQueryCount++;
+            printProgress();
+        }
+
+        producerTask.stopProducingEvents();
+    }
+
+    @Override
+    protected void teardown(Throwable t) {
+        ingestionExecutorService.shutdown();
+        try {
+            if (!ingestionExecutorService.awaitTermination(10, TimeUnit.SECONDS)) {
+                ingestionExecutorService.shutdownNow();
+            }
+        } catch (InterruptedException ex) {
+            ingestionExecutorService.shutdownNow();
+            Thread.currentThread().interrupt();
+        }
+    }
+
+    static class DataIngestionTask implements Runnable {
+        private final SqlService sqlService;
+        private final String sourceName;
+        private final long queryTimeout;
+        private final AtomicBoolean continueProducing;
+
+        DataIngestionTask(
+                SqlService sqlService,
+                String sourceName,
+                long queryTimeout) {
+            this.sqlService = sqlService;
+            this.sourceName = sourceName;
+            this.queryTimeout = queryTimeout;
+            this.continueProducing = new AtomicBoolean(true);
+        }
+
+        @Override
+        public void run() {
+            long startTime = System.currentTimeMillis();
+            AtomicInteger currentEventStartTime = new AtomicInteger(EVENTS_START_TIME);
+            AtomicInteger currentEventEndTime = new AtomicInteger();
+
+            while (continueProducing.get()) {
+                // ingest data to Kafka using new timestamps
+                currentEventStartTime.set(currentEventStartTime.get() + EVENTS_COUNT_PER_BATCH);
+                currentEventEndTime.set(currentEventStartTime.get() + EVENTS_COUNT_PER_BATCH);
+
+                String sql = "INSERT INTO " + sourceName + " VALUES" +
+                        TestRecordProducer.produceTradeRecords(
+                                currentEventStartTime.get(),
+                                currentEventEndTime.get(),
+                                EVENT_TIME_INTERVAL);
+
+                try (SqlResult res = sqlService.execute(sql)) {
+                    AbstractSoakTest.assertEquals(0L, res.updateCount());
+                } catch (HazelcastSqlException e) {
+                    continue;
```

Comment:
The intent here is to prevent the `run()` method from throwing an exception when the `DataIngestionTask` is told to shutdown. This can happen if the `teardown()` method is called while this thread is waiting on `sqlService.excecute()`. This doesn't cause the test to actually fail, but it just gives an ugly stacktrace on the client. I'm not sure if we care about any other failures that might occur here either. If there is an execution error sending a batch, it will sleep for the timeout period and send the next one. We only care about the streaming query in the main thread. The data ingestion need not be implemented using HZ SQL API, but could be done with any method that publishes to Kafka topic(s).

With the above in mind, do you have any suggestion for how to handle these situations? I'm thinking that at least logging the exception would be an improvement, rather than just silently continue. Wdyt?

Reply:
My main point is following - what happens if `sqlService.execute(sql)` will throw `HazelcastSqlException` for every command (for whatever reason). You can simulated for example by making SQL command incorrect (for example change `INSERT` to `INSERTINCORRECT`). Then this will run in cycle indefinitely. And since there's no check about a progress in the test then we will not know about it until it will fail at the end of soak tests run, right? It means that if we will do common 2 days soak test run and it will fail in the first minute we still will not know about it until the test is finished after 2 days.

Another thing which should be improved is that if it throws the exception (which also means that the command was not processed properly, i.e. data wasn't added) then we should rerun it with the same data, otherwise some data which should be inserted will be missing.

My general point for soak tests is - it's good to try how they behave if something goes wrong - will we know that it failed earlier than after 2 days? Do we have reasonable exception? etc.